igvc_navigation/src/mapper/octomapper.cpp

```diff
@@ -0,0 +1,215 @@
+#include <omp.h>
+#include <unordered_set>
+
+#include <octomap/octomap.h>
+#include <octomap_ros/conversions.h>
+
+#include "octomapper.h"
+
+using radians = double;
+Octomapper::Octomapper(ros::NodeHandle pNh)
+{
+  igvc::getParam(pNh, "octree/resolution", octree_options_.resolution);
+
+  igvc::getParam(pNh, "octree/clamping/max", octree_options_.max);
+  igvc::getParam(pNh, "octree/clamping/min", octree_options_.min);
+
+  igvc::getParam(pNh, "map/length", map_options_.length);
+  igvc::getParam(pNh, "map/width", map_options_.width);
+  igvc::getParam(pNh, "map/log_odds_default", map_options_.default_logodds);
+  std::string map_encoding;
+  igvc::getParam(pNh, "map/encoding", map_encoding);
+
+  map_options_.resolution = octree_options_.resolution;
+
+  // If we wanted more precision
+  if (map_encoding == "CV_8UC1")
+  {
+    map_encoding_ = CV_8UC1;
+  }
+  else
+  {
+    map_encoding_ = CV_8UC1;
+  }
+}
+
+void Octomapper::create_octree(pc_map_pair &pair) const
+{
+  pair.octree = boost::make_shared<octomap::OcTree>(octree_options_.resolution);
+  pair.octree->setClampingThresMin(octree_options_.min);
+  pair.octree->setClampingThresMax(octree_options_.max);
+  pair.octree->enableChangeDetection(true);
+  octomap::point3d min(-map_options_.length / 2.0, -map_options_.width / 2.0, -1);
+  octomap::point3d max(map_options_.length / 2.0, map_options_.width / 2.0, -1);
+  pair.octree->setBBXMin(min);
+  pair.octree->setBBXMax(max);
+}
+
+void PCL_to_Octomap(const pcl::PointCloud<pcl::PointXYZ> &pcl, octomap::Pointcloud &octo)
+{
+  sensor_msgs::PointCloud2 pc2;
+  pcl::toROSMsg(pcl, pc2);
+  octomap::pointCloud2ToOctomap(pc2, octo);
+}
+
+uchar toCharProb(float p)
+{
+  return static_cast<uchar>(p * 255);
+}
+
+float fromLogOdds(float log_odds)
+{
+  return 1 - (1 / (1 + exp(log_odds)));
+}
+
+std::pair<int, int> Octomapper::toMapCoordinates(double x, double y) const
+{
+  int x_map = static_cast<int>((map_options_.length / 2.0 + x) / octree_options_.resolution);
+  int y_map = static_cast<int>((map_options_.width / 2.0 + y) / octree_options_.resolution);
+  return std::make_pair(x_map, y_map);
+}
+
+void Octomapper::get_updated_map(struct pc_map_pair &pc_map_pair) const
+{
+  if (pc_map_pair.map == nullptr)
+  {
+    create_map(pc_map_pair);
+  }
+
+  // Traverse entire tree
+  std::vector<std::vector<float>> odds_sum(
+      static_cast<unsigned long>(map_options_.lengthGrid()),
+      std::vector<float>(static_cast<unsigned long>(map_options_.widthGrid()), map_options_.default_logodds));
+  for (octomap::OcTree::iterator it = pc_map_pair.octree->begin(), end = pc_map_pair.octree->end(); it != end; ++it)
+  {
+    // If this is a leaf at max depth, then only update that node
+    if (it.getDepth() == pc_map_pair.octree->getTreeDepth())
+    {
+      auto [x, y] = toMapCoordinates(it.getX(), it.getY());
```

Comment:
I think your current implementation of toMapCoordinates returns incorrect `x` and `y` coordinates here, as they are offset from the center of the map, not necessarily the designated map start location

Reply:
fixed.